File: frontend/src/domain/vod/value-objects/vod-url.ts
/**
 * VOD URL value object with validation.
 */

import { ValidationError } from '../../shared/errors';
import { Platform } from './platform';

export class VodUrl {
    private constructor(
        readonly url: string,
        readonly platform: Platform
    ) {
        this.validate();
    }

    private validate(): void {
        if (!this.url) {
            throw new ValidationError('vod_url', this.url, 'URL cannot be empty');
        }

        if (this.platform === Platform.Twitch) {
            const twitchPattern = /https?:\/\/(www\.)?twitch\.tv\/videos\/\d+/;
            if (!twitchPattern.test(this.url)) {
                throw new ValidationError(
                    'vod_url',
                    this.url,
                    'Invalid Twitch URL format. Expected: twitch.tv/videos/[id]'
                );
            }
        } else if (this.platform === Platform.YouTube) {
            const youtubePattern =
                /https?:\/\/(www\.)?(youtube\.com\/watch\?v=|youtu\.be\/)[\w-]+/;
            if (!youtubePattern.test(this.url)) {
                throw new ValidationError(
                    'vod_url',
                    this.url,
                    'Invalid YouTube URL format. Expected: youtube.com/watch?v=[id]'
                );
            }
        }
    }

    static create(url: string, platform: Platform): VodUrl {
        return new VodUrl(url, platform);
    }

    static fromString(url: string): VodUrl {
        if (url.includes('twitch.tv')) {
            return new VodUrl(url, Platform.Twitch);
        } else if (url.includes('youtube.com') || url.includes('youtu.be')) {
            return new VodUrl(url, Platform.YouTube);
        } else {
            throw new ValidationError(
                'vod_url',
                url,
                'Cannot detect platform from URL'
            );
        }
    }

    extractId(): string {
        if (this.platform === Platform.Twitch) {
            const match = this.url.match(/\/videos\/(\d+)/);
            return match?.[1] ?? '';
        } else if (this.platform === Platform.YouTube) {
            const match = this.url.match(/(?:v=|youtu\.be\/)([\w-]+)/);
            return match?.[1] ?? '';
        }
        return '';
    }

    equals(other: VodUrl): boolean {
        return this.url === other.url && this.platform === other.platform;
    }
}

File: backend/streamcraft/domain/vod/value_objects/vod_url.py
"""VOD URL value object with validation."""

import re
from dataclasses import dataclass

from streamcraft.domain.vod.errors.vod_errors import InvalidVodUrlError
from streamcraft.domain.vod.value_objects.platform import Platform


@dataclass(frozen=True, slots=True)
class VodUrl:
    """Validated VOD URL."""

    url: str
    platform: Platform

    def __post_init__(self) -> None:
        """Validate URL."""
        if not self.url:
            raise InvalidVodUrlError(self.url, "URL cannot be empty")

        if self.platform == Platform.TWITCH:
            if not re.match(r"https?://(www\.)?twitch\.tv/videos/\d+", self.url):
                raise InvalidVodUrlError(
                    self.url, "Invalid Twitch URL format. Expected: twitch.tv/videos/[id]"
                )
        elif self.platform == Platform.YOUTUBE:
            if not re.match(
                r"https?://(www\.)?(youtube\.com/watch\?v=|youtu\.be/)[\w-]+", self.url
            ):
                raise InvalidVodUrlError(
                    self.url, "Invalid YouTube URL format. Expected: youtube.com/watch?v=[id]"
                )

    @classmethod
    def from_string(cls, url: str) -> "VodUrl":
        """Create VodUrl from string, auto-detecting platform."""
        if "twitch.tv" in url:
            return cls(url=url, platform=Platform.TWITCH)
        elif "youtube.com" in url or "youtu.be" in url:
            return cls(url=url, platform=Platform.YOUTUBE)
        else:
            raise InvalidVodUrlError(url, "Cannot detect platform from URL")

    def extract_id(self) -> str:
        """Extract the VOD ID from the URL."""
        if self.platform == Platform.TWITCH:
            match = re.search(r"/videos/(\d+)", self.url)
            return match.group(1) if match else ""
        elif self.platform == Platform.YOUTUBE:
            match = re.search(r"(?:v=|youtu\.be/)([\w-]+)", self.url)
            return match.group(1) if match else ""
        return ""
